### PR TITLE
[MNG-8031] Backport concurrent friendly transport listener

### PR DESCRIPTION
**Pull request by cstamas**
Backport of simplex transfer listener.

---

https://issues.apache.org/jira/browse/MNG-8031

### CODE REVIEW

**Review by slawekjaranowski — APPROVED** [review]

File: maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link TransferListener} implementation that wraps another delegate {@link TransferListener} but makes it run
+ * on single thread, keeping the listener logic simple. This listener also blocks on last transfer event to allow
+ * output to perform possible cleanup. It spawns a daemon thread to consume queued events that may fall in even
+ * concurrently.
+ *
+ * @since 4.0.0
```
Comment (CrazyHZM) [inline maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java:44]: If we support it in 3.9.7, should it be changed to 3.9.7?

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java (hunk 1 above)
Comment (cstamas) [inline maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java:44]:
```suggestion
 * @since 3.9.7
```

**Review by CrazyHZM — APPROVED** [review]

File: maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final ConcurrentHashMap<File, Boolean> ongoing = new ConcurrentHashMap<>();
+
+    @Override
+    public void transferInitiated(TransferEvent event) {
+        ongoing.putIfAbsent(event.getResource().getFile(), Boolean.TRUE);
```
Comment (pshevche) [inline maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java:157]: @cstamas , `TransferResource` supports `file` being `null` ([ref](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/transfer/TransferResource.java#L166)). Althought the field is deprecated, the replacement is also nullable ([ref](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/transfer/TransferResource.java#L178)). However, this listener implementation will break if the resource has no `file` set. Was it a conscious decision?

Comment (cstamas) [reply maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java:157]: No, this is a probably oversight... care to provide a PR for this? And thanks!

Comment (pshevche) [reply maven-embedder/src/main/java/org/apache/maven/cli/transfer/SimplexTransferListener.java:157]: Yeah, I can do that. I'll look into it in https://issues.apache.org/jira/browse/MNG-8150

**Pull request merged**